File: top-interview-questions-easy/tree/LeetCode101.cc
#include<iostream>
using namespace std;

struct TreeNode{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr){}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        if(root == nullptr){
            return true;
        }
        return checkSymmetric(root->left,root->right);
    }

    bool checkSymmetric(TreeNode* left,TreeNode* right){
        if(left == nullptr && right == nullptr){
            return true;
        }
        if(left == nullptr || right == nullptr){
            return false;
        }
        if(left->val != right->val){
            return false;
        }
        else return (checkSymmetric(left->left,right->right) && checkSymmetric(left->right,right->left));
    }
};

File: top-interview-questions-easy/Tree/LeetCode101.go
package Tree

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func isSymmetric(root *TreeNode) bool {
	return isMirrot(root, root)
}

func isMirrot(tree1, tree2 *TreeNode) bool {

	//定义边界条件
	if tree1 == nil && tree2 == nil {
		return true
	}
	if tree1 == nil || tree2 == nil {
		return false
	}

	return (tree1.Val == tree2.Val) && isMirrot(tree1.Left, tree2.Right) && isMirrot(tree1.Right, tree2.Left)
}
